dali_tf_plugin/dali_dataset_op.cc

```diff
@@ -292,45 +534,120 @@ class DALIDatasetOp::Dataset::Iterator : public DatasetIterator<Dataset> {
+    return Status::OK();
+  }
 
+  /**
+   * @brief Check if samples have the same ndim, dtype etc.
+   *
+   * This probably is already handled by TF, as dataset probably can't
+   * dynamically change its shape. And we can probably check the declared output
+   * structure in Python level.
+   *
+   * TODO(klecki): add those checks in Python for clear errors
+   */
+  Status VerifyUniform(const Batch &input_batch, int input_idx) {
+    if (input_batch.empty()) {
+      return errors::InvalidArgument("Empty batch for input: ", input_idx, ".");
+    }
+    int ndim = input_batch[0].dims();
+    auto dtype = input_batch[0].dtype();
+    for (auto &sample : input_batch) {
+      if (sample.dims() != ndim) {
+        return errors::InvalidArgument(
+            "Inconsistent dimensionality of samples in a batch for input: ", input_idx,
+            ", got sample with: ", sample.dims(), " dimensions while the first one has: ", ndim,
+            " dimensions.");
+      }
+      if (sample.dtype() != dtype) {
+        return errors::InvalidArgument("Inconsistent dtype of samples in a batch for input: ",
+                                       input_idx, ", got sample with: ", sample.dtype(),
+                                       " dtype while the first one has: ", dtype, " dtype.");
+      }
+    }
     return Status::OK();
   }
 
-  ~Iterator() {
-    if (enable_memory_stats_) {
-      size_t N;
-      daliExecutorMetadata *meta;
-      daliGetExecutorMetadata(&pipeline_handle_, &meta, &N);
-      std::cout << "DALI operator memory statistics: " << std::endl;
-      for (size_t i = 0; i < N; ++i) {
-        std::cout << "Operator " << meta[i].operator_name;
-        for (size_t j = 0; j < meta[i].out_num; ++j) {
-          std::cout << "   output [ " << j << " ] : " << meta[i].real_size[j] << "B allocated "
-                    << meta[i].max_real_size[j] << "B max allocated " << meta[i].reserved[j]
-                    << "B reserved" << meta[i].max_reserved[j] << "B max reserved";
-          if (j != meta[i].out_num - 1) {
-            std::cout << ",";
-          }
-        }
-        std::cout << std::endl;
+
+  /**
+   * @brief Helper function that repacks the `Batch` (which is an list of samples returned by
+   * GetNext()), to the format used by DALI C API for feeding External Source.
+   *
+   * Outputs: ptrs, dtype, shapes, ndim
+   * Inputs: input_batch
+   */
+  Status RepackNonContiguousBatch(std::vector<const void *> &ptrs, dali_data_type_t &dtype,
+                                  std::vector<int64_t> &shapes, int64_t &ndim,
+                                  const Batch &input_batch) {
+    int batch_size = dataset()->pipeline_def_.batch_size;
+    assert(input_batch.size() > 0);
+    assert(input_batch.size() == batch_size);
+
+    ptrs.resize(batch_size, nullptr);
+    dtype = TfToDaliType(input_batch[0].dtype());
+    ndim = input_batch[0].dims();
+    shapes.reserve(batch_size * ndim);
+    shapes.clear();
```

Comment:
```suggestion
    shapes.clear();
    shapes.reserve(batch_size * ndim);
```
nitpick: Looks more natural to me. Any reason to have this order?

Reply:
Nope. Just the way I wrote it.